### DISCUSSION TITLE
CardNumber: No toma los settings en el modulo field(cardNumber) via Security Fields

### DISCUSSION THREAD

**Discussion by dantovbein** [mercadopago/sdk-js#146] [General]
Que tal,
Estoy intentando cambiar la configuración del `cardNumber`, precisamente el `length` de 8 a 16 dígitos, basándome en la [siguiente documentación](https://github.com/mercadopago/sdk-js/blob/main/API/fields.md#field-instanceupdateproperties) y lamentablemente no estoy viendo reflejados los settings.

Este es mi código:
```
const cardNumberField = mercadoPago.fields
          .create('cardNumber', {
            placeholder: '4078 2625 1269 9821',
          })
          .on('ready',onReadyMPSecureFields)
          .on('binChange', onBinChange)
          .update({
            settings: {
              length: 16,
            }
          })
          .mount('form-checkout__cardNumber')
```
También probé de la siguiente forma:
```
const cardNumberField = mercadoPago.fields
          .create('cardNumber', {
            placeholder: '4078 2625 1269 9821',
          })
          .on('ready',onReadyMPSecureFields)
          .on('binChange', onBinChange)
          .mount('form-checkout__cardNumber')

cardNumberField
.update({
            settings: {
              length: 16,
            }
          })
```

Necesito que si o si sea de 16 dígitos porque esta información la guardo en una API externa que si recibe menos de 16 dígitos me devuelve un error.

Agradecería la pronta respuesta 🙏

**Comment by lucmkz** [comment]
Buenas tardes, @dantovbein dantovbein todo bien?
Gracias por contactarnos.

Esta configuración solo se puede establecer después de recibir el _BIN_ de la tarjeta (los primeros ocho dígitos de la tarjeta)

Entonces creo que para cumplir con su flujo, el update debe realizarse después de que se devuelva el número de tarjeta.

Funciona así porque para mejorar nuestras validaciones, cuando el bin aún no existe, el número de caracteres posibles en el input del número de tarjeta es de 24 (el tamaño máximo posible de la tarjeta de crédito). De todos modos, mejoraremos la documentación para incluir este punto.

Es importante recordar que luego de llenar el bin, al llamar a el method `getPaymentMethods`, se devuelve toda esta información (número de caracteres, CVV obligatorio, entre otros...) todo dentro del objeto **_results.settings_**, sería interesante establecer este retorno como la cantidad de caracteres porque hay tarjetas de por ejemplo solo 13 dígitos que son válidas, y esto sin duda puede aumentar su conversión.

Este es aquí el formato que devuelve los datos.

```
[
  {
    "id": "visa",
    "name": "Visa",
    "payment_type_id": "credit_card",
    "status": "active",
    "secure_thumbnail": "https://www.mercadopago.com/org-img/MP3/API/logos/visa.gif",
    "thumbnail": "http://img.mlstatic.com/org-img/MP3/API/logos/visa.gif",
    "deferred_capture": "supported",
    "settings": {
      "bin": {
        "pattern": "^(4)",
        "exclusion_pattern": "^(400163|400176|400178|400185|400199|423808|439267|471233|473200|476332|482481|451416|438935|(40117[8-9])|(45763[1-2])|457393|431274)",
        "installments_pattern": "^(?!(417401|453998|426398|462437|451212|456188))"
      },
      "card_number": {
        "length": 16,
        "validation": "standard"
      },
      "security_code": {
        "mode": "mandatory",
        "length": 3,
        "card_location": "back"
      }
    },
    "additional_info_needed": [
      {}
    ],
    "min_allowed_amount": 0.5,
    "max_allowed_amount": 60000,
    "accreditation_time": 2880,
    "financial_institutions": {},
    "processing_modes": "aggregator"
  }
]
```

según la documentación
https://www.mercadopago.com.ar/developers/pt/reference/payment_methods/_payment_methods/get

**Comment by dantovbein** [comment]
@lucmkz , muchas gracias por el soporte.

Te referis a llamar el método `update` en el `binChange` cuando `bin` es `!==null`?

**Comment by lucmkz** [comment]
Buenas tardes @dantovbein!

 Si exacto, aqui en esta seccion de la [documentacion](https://www.mercadopago.com.ar/developers/es/docs/checkout-api/integration-configuration/card/integrate-via-core-methods#bookmark_obtener_m%C3%A9todos_de_pago_de_la_tarjeta) se despliega todo lo que debes hacer despues de recibir el bin, es decir `bin` es `!==null`, nota que entre los metodos esta `updatePCIFieldsSettings`, donde hay un ejemplo de esta verificacion `if (bin && bin !== currentBin){...}`

**Comment by dantovbein** [comment]
Ok!
@lucmkz Lo que pude hacer es actualizar los `settings` pero el `binChange` se sigue invocando cuando los digitos son 8 en vez de 16 como yo preciso.

1. ¿Hay alguna manera de obtener el número de tarjeta completo en vez del `bin`?
2. ¿Hay alguna forma de definirle un valor al `input` `cardNumber` que está dentro del `iframe`?
Gracias

**Comment by lucmkz** [comment]
No se si entendí bien tu pregunta, pero te respondo lo que creo que es tu duda jeje

1) es posible recuperar todos los datos de la tarjeta, desde el momento en que hay más de ocho dígitos (bin) en el input, es decir, de ocho en adelante, las validaciones se producirán correctamente desde el momento en que solo hay siete dígitos en el input, estas validaciones dejan de existir y los inputs vuelven al estado inicial.

2) por razones de seguridad, no es posible tratar directamente ningún dato sensible (número de tarjeta, código de seguridad y fecha de caducidad) de los IFrames. Por lo que no es posible establecer el valor del número de tarjeta.
